designs/2020-es-module-support/README.md

```diff
@@ -0,0 +1,159 @@
+│ (CJS entry point) ├──BUILD_STEP──▶│   (UMD bundle)  │     CJS ---▶   "main": "espress.js",
+│                   │               │                 │
+└-------------------┘               └-----------------┘
+```
+
+
+The strategy that probably makes the most sense in this initial work is to do what is known as a "dual package" solution, where
+espree provides an ESM, and a CJS module.
+
+Providing a dual package shouldn't affect any existing espree users, but provies an ESM option for users that want it. 
+
+
+proposed build process in this RFC:
+```
+                                  ┌-------------------┐     
+                                  │                   │       
+                                  │  dist/espree.cjs  │     
+                                  │ (CJS entry point) │                package.json: 
+┌-------------------┐             │                   │     
+│                   │             └--▲----------------┘     CJS ---▶   "main": "dist/espress.cjs",  
+│     espree.js     │                │                                 "exports": {
+│ (ESM entry point) ├───BUILD_STEP───┤                      ESM ---▶     "import": "espree.js",
+│                   │                │                      CJS ---▶     "require": "./dist/espree.cjs" 
+└-------------------┘                ▼                                 },
+                                  ┌----------------┐                   "type": "module"
+                                  │                │
+                                  │ dist/espree.js │
+                                  │  (UMD bundle)  │
+                                  │                │
+                                  └----------------┘
+```
+
+browserify is specifically oriented around commonjs as the input format, so I also propose replacing it with rollup, which understands and expects ESM as it's input format.
+
+
+## Documentation
+
+The changes should be described in the migration guide for whichever major version of espree this goes into (currently considering the upcoming 8.x line.)
+
+Despite the fact that we don't expect this to break the existing API, it probably makes sense to formally announce this via blog post as well, because it is a somewhat dramatic change.
+
+
+## Drawbacks
+
+The javascript build/tooling ecosystem is already monstrously complex, and this change adds another output format, which further complicates things (compare the 2 graphs above to see this additional complexity visualized.) 
+
+I do believe this is a temporary scenario; as time goes on, the UMD and CJS entry points could be dropped altogether, along with the build step. But that will take months, maybe even years depending on how comfortable the community is with this change and how widely adopted the ESM format becomes.
+
+
+## Backwards Compatibility Analysis
+
+The dual package approach _should_ not break espree for anyone currently using espree.
```

Comment:
I believe there will be no difference for Node apps/libs that `require("espree")`, but there are some changes that might break someone else:

* We are changing the `import` API: certainly when a module that `import`s espree is run in Node (they'll have to switch from `import espree from "espree"` to `import * as espree from "espree"`) and _maybe_ when the imports are interpreted by some tools (e.g., bundlers) depending on how the tool works.
* Adding `exports` to `package.json` restricts modules that can be imported/required from outside.

Reply:
I don't think we'd need to use `import * as espree from "espree"` based on the latest Node.js compat.

The second point is valid. Just adding `exports` to `package.json` is a breaking change.